scheduler/stack.go

```diff
@@ -70,7 +70,8 @@ type GenericStack struct {
 
 func (s *GenericStack) SetNodes(baseNodes []*structs.Node) {
 	// Shuffle base nodes
-	shuffleNodes(baseNodes)
+	idx, _ := s.ctx.State().LatestIndex()
```

Comment:
I can't believe how many words I'm writing about `we need a nonce` :sweat_smile:, however I think there are some subtle pros and cons to this approach:

**tl;dr** This allows writing reproducible tests against deterministic state stores which is the main improvement we're looking for! :shipit: 

Pros:

1. Subsequent scheduling attempts will be shuffled differently! This accomplishes that goal perfectly. :tada:
2. Test cases with deterministic state store updates will make shuffle nodes deterministically.

Cons:

1. When trying to reproduce scheduling behavior from a user's snapshot, the `LatestIndex()` will be the snapshot's index. The snapshot was likely taken long (in Raft index terms (pun not intended)) after when the scheduling attempt we wish to reproduce was made. Obviously in cases like this the entire state snapshot is likely divergent from the case we wish to inspect, so we can only _reduce_ non-determinism, not remove it. I'm unsure how much value there is in _reducing_ non-determinism in this case.
2. Workers only ever enforce the snapshots are **>=** a Raft index (Eval.WaitIndex initially, the RefreshIndex on subsequent attempts). Whether the index is `==` (reproducible shuffle!) or `>` (oh no) is entirely dependent on races between the Raft/fsm and scheduling subsystems. Even if the indexes are only off-by-1 the nodes will be shuffled completely differently.

Reply:
Given that the primary use case for this is to make it so that we can have deterministic tests, maybe the right way to deal with this is to just have the index used for the shuffle get logged somewhere. That way we can feed that index to the shuffle function as part of tests?

Reply:
Yeah that or the retry # option.

Although I don't think there's a rush as this should work for those tests where we either bypass raft and manually insert objects into the state store *or* we load a snapshot into an otherwise inactive server and perform some scheduling attempts.